### DISCUSSION TITLE
apply_filter returns empty dict?

### DISCUSSION THREAD

**Discussion by jdgre1** [cnr-isti-vclab/PyMeshLab#44] [Q&A] [answered]
Hello,
firstly thank-you for your help. 
I have installed pymeshlab on ubuntu via "pip3 install pymeshlab" and have loaded meshes for filtering options as seen below:

********************************************************************
```
ms = ml.MeshSet()
ms.load_new_mesh("../data/mesh_B.ply")
mesh = ms.current_mesh()
#print(mesh.vertex_matrix())
ms.load_new_mesh("../data/mesh_A.ply")
mesh = ms.current_mesh()
#print(mesh.vertex_matrix())
mesh_B_dist = ms.apply_filter('distance_from_reference_mesh', measuremesh = 0, refmesh = 1, signeddist = True, maxdist = 100)
convex_hullfilter = ms.apply_filter('convex_hull')
noisy_isosurface_filter = ms.apply_filter('noisy_isosurface', resolution=128)
print(noisy_isosurface_filter, convex_hullfilter, mesh_B_dist)
```

OUTPUT: `{} {} {}`
(Three empty python dictionaries)
********************************************************************
The meshes are not empty as printing the vertex_matrix indeed shows a numpy array. 

Am I doing something wrong? I am just wondering why all my filters are returning empty dictionaries..
I look forward to hearing back from you. Many thanks in advance.

**Comment by ZhuoshengKuang** [comment]
I am having the same issue, I am specifically interested in how to getting out the volume figures after applying filter ms.apply_filter('compute_geometric_measures').

Please help.
Thanks

**Comment by alemuntoni** [comment, marked as answer]
I'll try to improve the documentation on this point, but, as said in the [MeshSet documentation](https://pymeshlab.readthedocs.io/en/latest/classes/meshset.html):

> The MeshSet class represents a container that allows to collect a set of Meshes, where every Mesh corresponds to a layer of MeshLab. Every mesh has an unique ID that identifies it inside the MeshSet. From the MeshSet class, you can apply all the filters of MeshLab by calling the pmeshlab.MeshSet.apply_filter() method. [...]  **Sometimes, filters change the current mesh when applied, setting it to their output. The behaviour is always the same of MeshLab**.

Therefore, filters like `convex_hull` and `noisy_isosurface` do not return anything. They create a new mesh inside the MeshSet with their output (same behaviour as MeshLab).

Filters that in MeshLab print information in the log window, in PyMeshLab should put the same information in a dictionary that is returned as output, but just a few of them already do that (see #39). We are working on all the filters that should return some values.
In the meantime, everything that is shown in the MeshLab log (with a lot of other info) should be printed in PyMeshLab if the verbosity of the MeshSet is set to true:
```
ms.set_versbosity(True)
```

**Reply by jdgre1** [reply]
Oh... of course. Many thanks for your quick response: I realise that makes a lot of sense.. 
I can verify that it works perfectly. Thanks very much :)

**Reply by ZhuoshengKuang** [reply]
> I'll try to improve the documentation on this point, but, as said in the [MeshSet documentation](https://pymeshlab.readthedocs.io/en/latest/classes/meshset.html):
> 
> > The MeshSet class represents a container that allows to collect a set of Meshes, where every Mesh corresponds to a layer of MeshLab. Every mesh has an unique ID that identifies it inside the MeshSet. From the MeshSet class, you can apply all the filters of MeshLab by calling the pmeshlab.MeshSet.apply_filter() method. [...]  **Sometimes, filters change the current mesh when applied, setting it to their output. The behaviour is always the same of MeshLab**.
> 
> Therefore, filters like `convex_hull` and `noisy_isosurface` do not return anything. They create a new mesh inside the MeshSet with their output (same behaviour as MeshLab).
> 
> Filters that in MeshLab print information in the log window, in PyMeshLab should put the same information in a dictionary that is returned as output, but just a few of them already do that (see #39). We are working on all the filters that should return some values.
> In the meantime, everything that is shown in the MeshLab log (with a lot of other info) should be printed in PyMeshLab if the verbosity of the MeshSet is set to true:
> 
> ```
> ms.set_versbosity(True)
> ```

Thank you for your reply.
It is very dumb to ask this again.

After setting set_versbosity(True), it still return no geometric measure information.

ms = ml.MeshSet()
ms.set_versbosity(True)
ms.apply_filter('compute_geometric_measures')

it returns {}

Am I doing anything wrong here?

Thanks in advance.

**Reply by alemuntoni** [reply]
@ZhuoshengKuang did you loaded a mesh? :) it is missing that line in the snipped code you shared.
Also be sure to update pymeshlab to 0.1.8, since previous versions do not have this feature.
